src/storage/AgnosticStorage.ts

```diff
@@ -0,0 +1,170 @@
+import localforage from "localforage";
+
+/**
+ * AgnosticStorage provides a simple key-value storage interface.
+ * Completely wraps localStorage with memory caching for optimal performance.
+ */
+export class AgnosticStorage<T = any> {
+    private memoryCache: Map<string, T> = new Map();
+    private storage?: LocalForage;
+    private isPersistent: boolean;
+    private storeName: string;
+
+    /**
+     * Create a storage instance
+     * @param options.persist - Whether to enable persistence (default: false)
+     * @param options.storeName - Store name for persistence (default: 'default-store')
+     */
+    constructor(
+        options: {
+            persist?: boolean;
+            storeName?: string;
+        } = {}
+    ) {
+        this.isPersistent = options.persist ?? false;
+        this.storeName = options.storeName ?? "default-store";
+
+        if (this.isPersistent) {
+            this.storage = localforage.createInstance({
+                name: "state-channels-plus",
+                storeName: this.storeName,
+                driver: localforage.LOCALSTORAGE
+            });
+        }
+    }
+
+    /**
+     * Initialize the storage (loads existing data if persistent)
+     */
+    async initialize(): Promise<void> {
+        if (!this.isPersistent || !this.storage) {
+            return;
+        }
+
+        try {
+            const keys = await this.storage.keys();
+            for (const key of keys) {
+                const value = await this.storage.getItem<T>(key);
+                if (value !== null) {
+                    this.memoryCache.set(key, value);
+                }
+            }
+        } catch (error) {
+            console.warn("Failed to load existing data:", error);
+        }
+    }
+
+    async set(
+        key: string,
+        value: T,
+        options: { wait?: boolean } = {}
+    ): Promise<void> {
+        this.memoryCache.set(key, value);
+
+        if (this.isPersistent && this.storage) {
+            if (options.wait) {
+                await this.storage.setItem(key, value);
+            } else {
+                this.storage.setItem(key, value).catch((error) => {
+                    console.error(
+                        "Background persistence failed for key:",
+                        key,
+                        error
+                    );
+                });
+            }
+        }
+    }
+
+    async get(key: string): Promise<T | null> {
+        return this.memoryCache.get(key) ?? null;
```

Comment:
No need to have this function async, when the function body is fully sync and doesn't use async code